### PR TITLE
feat(blade): add PhoneNumber input

### PR DESCRIPTION
**Pull request by anuraghazra**
## Component Checklist

- [ ] Update Component Status Page
- [ ] Perform Manual Testing in Other Browsers
- [ ] Add KitchenSink Story
- [x] Add Interaction Tests (if applicable)
- [x] Add changeset

### CODE REVIEW

File: .github/workflows/blade-chromatic.yml [hunk 1]
```diff
@@ -18,7 +18,11 @@ jobs:
         uses: actions/setup-node@v3
         with:
           node-version: 18.12.1
+          # https://stackoverflow.com/a/73824182
+          registry-url: "https://npm.pkg.github.com"
```
Comment (saurabhdaware) [inline .github/workflows/blade-chromatic.yml:22]: what does this solve?

File: packages/blade/src/components/Dropdown/DropdownOverlay.web.tsx [hunk 2]
```diff
@@ -34,25 +34,30 @@ const _DropdownOverlay = ({
...
-    dropdownTriggerer !== dropdownComponentIds.triggers.AutoComplete;
+    dropdownTriggerer !== dropdownComponentIds.triggers.AutoComplete &&
+    referenceRef == undefined;
 
   const { refs, floatingStyles, context } = useFloating({
     open: isOpen,
     onOpenChange: setIsOpen,
     strategy: 'fixed',
-    placement: 'bottom-start',
+    placement: defaultPlacement,
```
Comment (saurabhdaware) [inline packages/blade/src/components/Dropdown/DropdownOverlay.web.tsx:53]: old unrelated change

**Review by saurabhdaware — COMMENTED** [review]
There's a lot of unrelated changes in diff. Can you pull master?

File: .github/workflows/blade-chromatic.yml (hunk 1 above)
Comment (anuraghazra) [reply .github/workflows/blade-chromatic.yml:22]: This was related to i18nify package's default registry being npm.pkg.github in our lockfile, without this + the NODE_AUTH_TOKEN the installation fails on CI

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    inputRef.current?.focus();
+  };
+
+  return (
+    <BaseInput
+      setInputWrapperRef={(node) => {
+        inputWrapperRef.current = node as HTMLInputElement;
+      }}
+      ref={mergedRef}
+      id="textinput"
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:148]: nit: make it phone-number-input? so we can differentiate with textinput

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+      autoFocus={autoFocus}
+      testID={testID}
+      size={size}
+      {...getKeyboardAndAutocompleteProps({
+        type: 'number',
+        keyboardReturnKeyType,
+        autoCompleteSuggestionType,
+        autoCapitalize,
+      })}
+      type="number"
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:206]: shouldn't it be `type="telephone"` 🤔

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+      {...styledProps}
+    />
+  );
+};
+
+const PhoneNumberInput = assignWithoutSideEffects(React.forwardRef(_PhoneNumberInput), {
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:212]: Add basic jsdoc?

File: packages/blade/src/components/Input/TextInput/TextInput.tsx [hunk 6]
```diff
@@ -129,88 +123,6 @@ type TextInputPropsWithLabel = {
 type TextInputProps = (TextInputPropsWithA11yLabel | TextInputPropsWithLabel) &
   TextInputCommonProps;
 
-const getKeyboardAndAutocompleteProps = ({
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/TextInput/TextInput.tsx:132]: Nice!

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 4 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:206]: Yup I think. Because of type="number" HTML's default number things are coming up like pressing up and down arrows increases and decreases the counter
image

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+};
+
+export const WithoutCountrySelector = PhoneNumberInputTemplate.bind({});
+WithoutCountrySelector.args = {
+  showCountrySelector: false,
+};
+
+export const WithoutDialCode = PhoneNumberInputTemplate.bind({});
+WithoutDialCode.args = {
+  showDialCode: false,
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.stories.tsx:51]: does this have valid usecase? looks a bit strange in UI 🤔

**Review by saurabhdaware — COMMENTED** [review]
UI looks good to me on storybook. Good job 🥳

File: packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    <Dropdown isOpen={isDropdownOpen} onOpenChange={setIsDropdownOpen}>
+      <DropdownButton
+        isDisabled={isDisabled}
+        size={size === 'medium' ? 'xsmall' : 'medium'}
+        variant="tertiary"
+        accessibilityLabel={countryNameFormatter.of(selectedCountry)}
+        icon={isDropdownOpen ? ChevronUpIcon : ChevronDownIcon}
+        iconPosition="right"
+      >
+        {/* @ts-expect-error */}
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx:80]: Don't want to expose React.ReactNode as type of the children for public component.

File: packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+          alt=""
+        />
+      </DropdownButton>
+      {isMobile ? (
+        <BottomSheet>
+          <BottomSheetHeader title="Sort By" />
+          <BottomSheetBody>{actionList}</BottomSheetBody>
+        </BottomSheet>
+      ) : (
+        <DropdownOverlay referenceRef={inputWrapperRef}>{actionList}</DropdownOverlay>
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx:95]: Makes the dropdown same width as the BaseInput and positions it properly.

File: packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        icon={isDropdownOpen ? ChevronUpIcon : ChevronDownIcon}
+        iconPosition="right"
+      >
+        {/* @ts-expect-error */}
+        <img
+          loading="lazy"
+          role="presentation"
+          width={flagSize[size]}
+          src={getFlagOfCountry(selectedCountry)['4X3']}
+          alt=""
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx:86]: Nope, role="presentation" is there. won't be announced either ways

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        isDisabled={isDisabled}
+        accessibilityLabel="Clear Input Content"
+      />
+    );
+  };
+
+  const flags = React.useMemo(() => getFlagsForAllCountries(), []);
+  const countryData = React.useMemo(() => {
+    if (countries) {
+      return countries.map((countryCode) => {
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:95]: Advantage of not sorting consumer passed array is that we don't need to expose another prop like [`preferedCountries`](https://react-international-phone.vercel.app/docs/Usage/PhoneInput#preferredcountries), if say user want's malaysia to be on top of the list they can just put MY on top of the passed array.

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+  };
+
+  return (
+    <BaseInput
+      setInputWrapperRef={(node) => {
+        inputWrapperRef.current = node as HTMLInputElement;
+      }}
+      ref={mergedRef}
+      id="textinput"
+      componentName="PhoneNumberInput"
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:149]: fixed

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 3 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:148]: Weird, why do we even need this? even on TextInput.

Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:148]: Done.

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 5 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:212]: done

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.stories.tsx (hunk 7 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.stories.tsx:51]: design mein people can remove the prefix text. Not sure about usecase.

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 4 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:206]: `telephone` doesn't do any validation or prevents other chars, users can type random characters too. 

If we use telephone we will manually have to do validation or prevent this behaviour. Let's keep it number for now if we find any issues we can later switch to telephone.

File: packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Input/PhoneNumberInput/CountrySelector.web.tsx:80]: oh yes yes I meant adding it in comment otherwise we wouldn't remember later 🙈

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:96]: makes sense

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 4 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx]: That is the expected call we had taken for phone numbers no? Article - https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/

Anyway I don't see any case where consumers won't do any validation from their end and since we have validations coming from i18n now, the other characters would be caught in that validation anyways.

Its also rare that someone would try to enter characters because visually it is clear that its phone number input


People might end up accidentally changing the phone number while scrolling because of the default behaviour of type number

https://github.com/razorpay/blade/assets/30949385/d7eab2cc-deb4-46ee-a630-2d10adc23a3f

Comment (saurabhdaware) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx]: I don't see in the latest preview btw so are we going with telephone?

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+      autoFocus={autoFocus}
+      testID={testID}
+      size={size}
+      {...getKeyboardAndAutocompleteProps({
+        type: 'number',
+        keyboardReturnKeyType,
+        autoCompleteSuggestionType,
+        autoCapitalize,
+      })}
+      type="telephone"
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:207]: are we going with this then?

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+};
+
+/**
+ * PhoneNumberInput is a component that allows users to input phone numbers.
+ * It provides a country selector dropdown to select the country code.
+ *
+ * @example
+ *
+ * ```ts
+ * <PhoneNumberInput />
```
Comment (saurabhdaware) [inline packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:220]: should we add basic validation example here? since pretty much every usage would require it along with validation?

**Review by saurabhdaware — APPROVED** [review]
🚢 🇮🇹

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 14 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:220]: Validation usecase is already added in storybook example.

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 13 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:207]: Oops, lol I was testing it out.

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 4 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx:207]: Fine with either one. We can change it to telephone.

File: packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx (hunk 13 above)
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/PhoneNumberInput.web.tsx]: yeah

**Review by saurabhdaware — DISMISSED** [review]

File: packages/blade/src/components/Input/PhoneNumberInput/types.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+  | 'onFocus'
+  | 'onBlur'
+  | 'onClick'
+  | 'leadingIcon'
+  | 'trailingIcon'
+  | 'accessibilityLabel'
+  | 'autoFocus'
+  | 'testID'
+  | 'keyboardReturnKeyType'
+  | 'autoCompleteSuggestionType'
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/types.ts:27]: Yeah, users can pass telephone as `autoCompleteSuggestionType`

File: packages/blade/src/components/Input/PhoneNumberInput/types.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  | 'isRequired'
+  | 'isDisabled'
+  | 'onFocus'
+  | 'onBlur'
+  | 'onClick'
+  | 'leadingIcon'
+  | 'trailingIcon'
+  | 'accessibilityLabel'
+  | 'autoFocus'
+  | 'testID'
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/types.ts:25]: done

File: packages/blade/src/components/Input/PhoneNumberInput/types.ts [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Callback that is called when the clear button is clicked.
+     */
+    onClearButtonClick?: () => void;
+    /**
+     * The size of the input field.
+     *
+     * @default medium
+     */
+    size?: 'medium' | 'large';
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/types.ts:102]: done

File: packages/blade/src/components/Button/BaseButton/BaseButton.tsx [hunk 18]
```diff
@@ -328,6 +329,8 @@ const _BaseButton: React.ForwardRefRenderFunction<BladeElementRef, BaseButtonPro
   const [isPressed, setIsPressed] = React.useState(false);
   const isLink = Boolean(href);
   const childrenString = getStringFromReactText(children);
+  const isChildrenComponent = childrenString === '[object Object]';
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Button/BaseButton/BaseButton.tsx:332]: Ideally should be childrenString.match('[object Object]'), so that it can work with multiple react childrens too but React.isValidElement will suffice for the country selector usecase. 

```
<Button>
  <p>hello</p>
  <p>world</p>
</Button>
```

File: packages/blade/src/components/Dropdown/__tests__/__snapshots__/Dropdown.web.test.tsx.snap [hunk 19]
```diff
@@ -96,7 +96,7 @@ exports[`<Dropdown /> should render dropdown and make it visible on click 1`] =
   -webkit-align-self: stretch;
   -ms-flex-item-align: stretch;
   align-self: stretch;
-  padding-right: 12px;
+  margin-right: 12px;
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Dropdown/__tests__/__snapshots__/Dropdown.web.test.tsx.snap]: Outdated snapshot

File: packages/blade/src/components/Dropdown/types.ts [hunk 20]
```diff
@@ -29,6 +29,7 @@ type DropdownOverlayProps = {
    */
   zIndex?: number;
   width?: BoxProps['width'];
+  referenceRef?: React.MutableRefObject<ContainerElementType | null>;
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Dropdown/types.ts:32]: "reference" refers to the floating UI's reference prop. 

https://floating-ui.com/docs/useFloating#setreference

I'll add a jsdoc comment for this.

File: packages/blade/src/components/Input/BaseInput/BaseInput.tsx [hunk 21]
```diff
@@ -157,7 +157,13 @@ type BaseInputCommonProps = FormInputLabelProps &
      *
      * eg: consumers can render a loader or they could render a clear button
      */
-    interactionElement?: ReactNode;
+    trailingInteractionElement?: ReactNode;
+    /**
+     * Element to be rendered before prefix. This is decided by the component which is extending BaseInput
+     *
+     * eg: consumers can render a loader or they could render a clear button
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/BaseInput/BaseInput.tsx:164]: fixed

File: packages/blade/src/components/Input/BaseInput/BaseInputVisuals.tsx [hunk 22]
```diff
@@ -198,7 +204,7 @@ export const BaseInputVisuals = ({
 
   const hasLeadingVisuals = hasLeadingIcon || hasPrefix;
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/BaseInput/BaseInputVisuals.tsx]: fixed

File: packages/blade/src/components/Input/BaseInput/baseInputStyles.ts [hunk 23]
```diff
@@ -211,11 +214,12 @@ export const getBaseInputStyles = ({
     paddingTop: makeSpace(theme.spacing[baseInputPaddingTokens.top[size]]),
     paddingBottom: makeSpace(theme.spacing[baseInputPaddingTokens.bottom[size]]),
     paddingLeft: makeSpace(
+      // TODO: hasInteractionElement: hasTrailingInteractionElement,
       getLeftPadding({ theme, isDropdownTrigger, hasLeadingIcon, hasPrefix, size }),
     ),
     paddingRight: getRightPadding({
       theme,
-      hasInteractionElement,
+      hasInteractionElement: hasTrailingInteractionElement,
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/BaseInput/baseInputStyles.ts:222]: Renamed. 

> Also leadingInteractionElement has no effect on padding?

Nope. Leading will not effect anything on "right" side

File: packages/blade/src/components/Input/BaseInput/baseInputStyles.ts [hunk 24]
```diff
@@ -211,11 +214,12 @@ export const getBaseInputStyles = ({
     paddingTop: makeSpace(theme.spacing[baseInputPaddingTokens.top[size]]),
     paddingBottom: makeSpace(theme.spacing[baseInputPaddingTokens.bottom[size]]),
     paddingLeft: makeSpace(
+      // TODO: hasInteractionElement: hasTrailingInteractionElement,
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/BaseInput/baseInputStyles.ts:217]: Yeah, removed this. We don't need this.

File: packages/blade/src/components/Input/PhoneNumberInput/types.ts [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Value of the input, Used to turn the input field to controlled so user can control the value
+     */
+    value?: string;
+    /**
+     * The default country code to be used in the input.
+     *
+     * @default "IN" or autodetect based on the user's locale
+     */
+    defaultCountryCode?: CountryCodeType;
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Input/PhoneNumberInput/types.ts:47]: Added, countryCode and onCountrySelection

**Comment by chaitanyadeorukhkar** [comment]
LGTM. Only 1 pending conversation going on slack regarding prop names. We can merge this once that is concluded.
